The `ui.data()` requires mutable borrow to `get_temp()`?

<!--
First look if there is already a similar bug report. If there is, upvote the issue with 👍

Please also check if the bug is still present in latest master! Do so by adding the following lines to your Cargo.toml:


[patch.crates-io]
egui = { git = "https://github.com/emilk/egui", branch = "master" }
# if you're using eframe:
eframe = { git = "https://github.com/emilk/egui", branch = "master" }
-->

**Describe the bug**
<!-- A clear and concise description of what the bug is. An image is good, a gif or movie is better! -->
It seems that inside the `ui.data()` lambda I can't `get_temp()` anything from the `IdTypeMap` because the reference to it passed to the lambda  apparently needs to be mutable. Why?

**To Reproduce**
Steps to reproduce the behavior:
1. Create an `eframe` project and plop down some widget
2. Inside the lambda that accepts `&mut egui::Ui` do
```rust
let x_id = ui.id().with("x");
let x = ui.data(|d| d.get_temp::<u32>(x_id).unwrap_or_default());
```
3.
```
error[E0596]: cannot borrow `*d` as mutable, as it is behind a `&` reference
  --> [DATA EXPUNGED].rs:41:33
   |
41 |             let x = ui.data(|d| d.get_temp::<u32>(x_id).unwrap_or_default());
   |                              -  ^^^^^^^^^^^^^^^^^^^^^^^ `d` is a `&` reference, so the data it refers to cannot be borrowed as mutable
   |                              |
   |                              consider changing this binding's type to be: `&mut IdTypeMap`

```

**Expected behavior**
<!-- A clear and concise description of what you expected to happen. -->
Since `get_temp()` clones the value, the `IdTypeMap` is not required to be mutable. Applies to all "read" methods.

**Screenshots**
<!-- If applicable, add screenshots to help explain your problem. -->

**Desktop (please complete the following information):**
 - OS: `Linux 5.18.0-14parrot1-amd64 #1 SMP PREEMPT_DYNAMIC Debian 5.18.14-1parrot1 (2022-08-07) x86_64 GNU/Linux`
 - Browser <!-- e.g. chrome, safari -->
 - Version `master` (7215fdfb7c7407b8085d53052582dac10124bdfc)

**Additional context**
<!-- Add any other context about the problem here. -->
Am I missing something? This seems very weird.

Possibly related:
- #2625

--- a/crates/egui/src/util/id_type_map.rs
+++ b/crates/egui/src/util/id_type_map.rs
@@ -366,10 +366,7 @@ impl IdTypeMap {
     #[inline]
     pub fn get_temp<T: 'static + Clone>(&self, id: Id) -> Option<T> {
         let hash = hash(TypeId::of::<T>(), id);
-        self.0
-            .get(&hash)
-            .and_then(|x| x.get_temp())
-            .cloned()
+        self.0.get(&hash).and_then(|x| x.get_temp()).cloned()
     }
 
     /// Read a value, optionally deserializing it if available.

--- a/crates/egui/src/util/id_type_map.rs
+++ b/crates/egui/src/util/id_type_map.rs
@@ -309,6 +309,7 @@ use crate::Id;
 ///
 /// Values are cloned when read, so keep them small and light.
 /// If you want to store something bigger, wrap them in `Arc<Mutex<…>>`.
+/// Also try `Arc<ArcSwap<…>>`.
 ///
 /// Values can either be "persisted" (serializable) or "temporary" (cleared when egui is shut down).
 ///

--- a/crates/egui/src/util/id_type_map.rs
+++ b/crates/egui/src/util/id_type_map.rs
@@ -373,6 +373,9 @@ impl IdTypeMap {
 
     /// Read a value, optionally deserializing it if available.
     ///
+    /// NOTE: A mutable `self` is needed because internally this deserializes on first call
+    /// and caches the result (caching requires self-mutability).
+    ///
     /// The call clones the value (if found), so make sure it is cheap to clone!
     #[inline]
     pub fn get_persisted<T: SerializableAny>(&mut self, id: Id) -> Option<T> {

--- a/crates/egui/src/util/id_type_map.rs
+++ b/crates/egui/src/util/id_type_map.rs
@@ -170,6 +170,14 @@ impl Element {
     }
 
     #[inline]
+    pub(crate) fn get_temp<T: 'static>(&self) -> Option<&T> {
+        match self {
+            Self::Value { value, .. } => value.downcast_ref(),
+            Self::Serialized { .. } => None,
+        }
+    }
+
+    #[inline]
     pub(crate) fn get_mut_temp<T: 'static>(&mut self) -> Option<&mut T> {
         match self {
             Self::Value { value, .. } => value.downcast_mut(),
@@ -355,11 +363,11 @@ impl IdTypeMap {
     ///
     /// The call clones the value (if found), so make sure it is cheap to clone!
     #[inline]
-    pub fn get_temp<T: 'static + Clone>(&mut self, id: Id) -> Option<T> {
+    pub fn get_temp<T: 'static + Clone>(&self, id: Id) -> Option<T> {
         let hash = hash(TypeId::of::<T>(), id);
         self.0
-            .get_mut(&hash)
-            .and_then(|x| x.get_mut_temp())
+            .get(&hash)
+            .and_then(|x| x.get_temp())
             .cloned()
     }
 

